src/guest_memory.rs

```diff
@@ -302,6 +306,96 @@ pub trait GuestMemoryMut {
     }
 }
 
+/// GuestMemorySnapshottable provides a way to retrieve a GuestMemory object.
+/// The vm-memory crate already provides trivial implementation for
+/// instances of GuestMemory (more precisely, for references to them
+/// or reference-counted GuestMemory objects), but the trait can also
+/// be implemented by a GuestMemoryMut object in order to obtain a
+/// consistent snapshot of the memory map.
+///
+/// In order to support generic mutable memory maps, devices (or other things
+/// that access memory) should store the memory as a GuestMemorySnapshottable<M>.
+/// In the following example, dummy structs that implement the GuestMemoryBackend trait
+/// describe how the VMM is storing the memory map.  We could also declare VirtioDevice
+/// as `<M: GuestMemory, S: GuestMemorySnapshottable<M>`, but this would lead
+/// to repetitive types such as `VirtioDevice<GuestMemoryMmap, &GuestMemoryMmap>`.
+///
+/// ```
+/// # #[cfg(feature = "backend-mmap")]
+/// # use vm_memory::GuestMemoryMmap;
+/// # use vm_memory::{GuestAddress, GuestMemory, GuestMemorySnapshottable};
+///
+/// # // Perhaps this could be added to vm-memory itself?
+/// trait GuestMemoryBackend<'a> {
+///   type Snapshot: GuestMemory;
+///   type Memory: 'a + GuestMemorySnapshottable<Self::Snapshot>;
+/// }
+///
+/// pub struct VirtioDevice<'a, MB: GuestMemoryBackend<'a>> {
+///     mem: Option<MB::Memory>,
+/// }
+///
+/// impl<'a, MB: GuestMemoryBackend<'a>> VirtioDevice<'a, MB> {
+///     fn new() -> Self {
+///         VirtioDevice { mem: None }
+///     }
+///     fn activate(&mut self, mem: MB::Memory) {
+///         self.mem = Some(mem)
+///     }
+/// }
+///
+/// # #[cfg(feature = "backend-mmap")]
+/// # fn get_mmap() -> GuestMemoryMmap {
+/// #     GuestMemoryMmap::new(&[(GuestAddress(0),0)]).unwrap()
+/// # }
+///
+/// # #[cfg(feature = "backend-mmap")]
+/// # fn test_1() {
+/// // Using `VirtioDevice` with an immutable GuestMemoryMmap:
+/// struct ImmutableMmap;
+/// impl<'a> GuestMemoryBackend<'a> for ImmutableMmap {
+///     type Snapshot = GuestMemoryMmap;
+///     type Memory = &'a GuestMemoryMmap;
+/// }
+/// let mut for_immutable_mmap: VirtioDevice<ImmutableMmap> =
+///     VirtioDevice::new();
+/// let mmap = get_mmap();
+/// for_immutable_mmap.activate(&mmap);
+/// # }
+/// ```
```

Comment:
In my original design, I used `M: GuestMemory` for both GuestMemoryBackend::Snapshot and GuestMemoryBackend::Memory.
The consideration is that:
1) for GuestMemory clients not caring about performance and only using stateless interfaces such as Bytes, it doesn't need to care about memory hotplug and snapshot().
2) for GuestMemory clients caring about performance or using stateful interfaces such as find_region/num_region/last_addr etc, it need to use the snapshot() method to get a stable GuestMemory snapshot.

With this proposed design, we need to teach all GuestMemory consumers about memory hotplug and snapshot(). But I'm on this side to clearly expose Memory/Snapshot concepts to vm-memory consumers.

And the changes to vm-virtio and virtio device drivers should be acceptable.

Reply:
> for GuestMemory clients not caring about performance and only using stateless interfaces such as Bytes, it doesn't need to care about memory hotplug and snapshot().

The main issue is that you could have a client caring about performance and potentially hotplug (`vm-virtio`), but it should not add any cost if the `GuestMemory` user does not need hotplug (like firecracker).

> we need to teach all GuestMemory consumers about memory hotplug and snapshot(). 

Not all of them. You can keep using `GuestMemory` just like now if you don't support hotplug. The need for changes only arises if you want to use `GuestMemoryMut` to support hotplug. Then you need to introduce `snapshot()`.

Reply:
> > for GuestMemory clients not caring about performance and only using stateless interfaces such as Bytes, it doesn't need to care about memory hotplug and snapshot().
> 
> The main issue is that you could have a client caring about performance and potentially hotplug (`vm-virtio`), but it should not add any cost if the `GuestMemory` user does not need hotplug (like firecracker).
There are two cases here,
1) vmm supports memory hotplug or not. If the vmm doesn't support hotplug, it's easy to handle, current GuestMemory trait just works without code changes.
2) components (such as device drivers) support memory hotplug or not when the vmm supports memory hotplug.
What I'm thinking is about case 2.

> 
> > we need to teach all GuestMemory consumers about memory hotplug and snapshot().
> 
> Not all of them. You can keep using `GuestMemory` just like now if you don't support hotplug. The need for changes only arises if you want to use `GuestMemoryMut` to support hotplug. Then you need to introduce `snapshot()`.

The issue arises when sharing code among VMMs. For VMM specific implementations, we could just choose one of hotplug or non-hotplug interfaces. To share code, such as vm-device/vm-virtio, we need to teach them about hotplug/snapshot().